rustworkx-core/src/connectivity/cycle_basis.rs

```diff
@@ -114,18 +181,140 @@ where
         graph_nodes = graph_nodes.difference(&temp_hashset).copied().collect();
         root_node = None;
     }
-    cycles
+    if edges {
+        EdgesOrNodes::Edges(cycles_edges)
+    } else {
+        EdgesOrNodes::Nodes(cycles_nodes)
+    }
+}
+
+/// Enum for custom return types of `cycle_basis()`.
+enum EdgesOrNodes<N, E> {
+    Nodes(Vec<Vec<N>>),
+    Edges(Vec<Vec<E>>),
+}
+/// Functions used to unwrap the desired datatype of `EdgesOrNodes`.
+impl<N, E> EdgesOrNodes<N, E> {
+    fn unwrap_nodes(self) -> Vec<Vec<N>> {
+        match self {
+            Self::Nodes(x) => x,
+            Self::Edges(_) => unreachable!(
+                "Function should only return instances of {}.",
+                std::any::type_name::<N>()
+            ),
+        }
+    }
+    fn unwrap_edges(self) -> Vec<Vec<E>> {
+        match self {
+            Self::Edges(x) => x,
+            Self::Nodes(_) => unreachable!(
+                "Function should only return instances of {}.",
+                std::any::type_name::<E>()
+            ),
+        }
+    }
+}
+
+/// Returns lists of `NodeIndex` representing cycles which form
+/// a basis for cycles of a given graph.
+///
+/// A basis for cycles of a graph is a minimal collection of
+/// cycles such that any cycle in the graph can be written
+/// as a sum of cycles in the basis.  Here summation of cycles
+/// is defined as the exclusive-or of the edges.
+///
+/// This is adapted from
+///    Paton, K. An algorithm for finding a fundamental set of
+///    cycles of a graph. Comm. ACM 12, 9 (Sept 1969), 514-518.
+///
+/// The function implicitly assumes that there are no parallel edges.
+/// It may produce incorrect/unexpected results if the input graph has
+/// parallel edges.
+///
+///
+/// Arguments:
+///
+/// * `graph` - The graph in which to find the basis.
+/// * `root` - Optional node index for starting the basis search. If not
+///     specified, an arbitrary node is chosen.
+///
+/// # Example
+/// ```rust
+/// use petgraph::prelude::*;
+/// use rustworkx_core::connectivity::cycle_basis;
+///
+/// let edge_list = [(0, 1), (0, 3), (0, 5), (1, 2), (2, 3), (3, 4), (4, 5)];
+/// let graph = UnGraph::<i32, i32>::from_edges(&edge_list);
+/// let mut res: Vec<Vec<NodeIndex>> = cycle_basis(&graph, Some(NodeIndex::new(0)));
+/// ```
+pub fn cycle_basis<G>(graph: G, root: Option<G::NodeId>) -> Vec<Vec<G::NodeId>>
+where
+    G: NodeCount,
+    G: IntoEdges,
+    G: IntoNodeIdentifiers,
+    G::NodeId: Eq + Hash,
+    G::EdgeId: Eq + Hash,
```

Comment:
Hmm, I'm mildly concerned about this change to the trait bounds here. It's also technically a backwards incompatible change because it's adding extra constraints (`G: IntoEdges` and `G::EdgeId: Eq + Hash`) on what is acceptable input for `graph`. I'm not sure how to fix this short of going against my earlier suggestion and re-duplicating the implementation and having 2 copies one for `cycle_basis` and the other for `cycle_basis_edges`. However, I'm also struggling to  come up with a real world example where an input graph type might not support `IntoEdges` given the previous traits. So I'm thinking we should just have an `upgrade` note documenting the potentially breaking API change to this function

Reply:
Interesting point. The trait IntoEdges is needed to access Edge information. I was unaware of the implications this change would bring. I will ass the `upgrade` note to the release notes accordingly.